src/main/java/frc/robot/subsystems/DriveBaseSubsystem.java

```diff
@@ -61,11 +65,43 @@ public DriveBaseSubsystem(Joystick joystick) {
     // differential drive
     m_differentialDrive = new DifferentialDrive(m_motorControllers[Constants.driveLeftFrontIndex].getSparkMax(), 
                                           m_motorControllers[Constants.driveRightFrontIndex].getSparkMax());
+
+    this.usingExternal = usingExternal; // gets key if using external or internal encoders
+
+    // external encoders            
+    m_extLeftEncoder = new Encoder(Constants.leftEncoderDIOone, Constants.leftEncoderDIOtwo, 
+    false, Encoder.EncodingType.k2X); // TODO: confirm correct configuration for encoder
+    m_extRightEncoder = new Encoder(Constants.rightEncoderDIOone, Constants.rightEncoderDIOtwo, 
+    false, Encoder.EncodingType.k2X);
+
+    m_extLeftEncoder.setReverseDirection(true);
+    
+    // internal encoders
+    m_internalLeftEncoder = m_motorControllers[Constants.kDriveLeftFrontIndex].getEncoder();
+    m_internalRightEncoder = m_motorControllers[Constants.kDriveRightFrontIndex].getEncoder();
+
+    m_internalLeftEncoder.setPositionConversionFactor(2 * Math.PI * Constants.wheelRadius / Constants.inchesInMeter); // calculate circumference then convert to meters
+    m_internalRightEncoder.setPositionConversionFactor(2 * Math.PI * Constants.wheelRadius / Constants.inchesInMeter);
+
   }
 
   @Override
   public void periodic() {
-    arcadeDrive();  // periodically runs the arcadeDrive function, avoid scheduling the command
+
+    // update odometry
+
+    // checks which encoders are being used: external or internal
+    if(usingExternal) {
+      m_odometry.update(
+        m_gyro.getRotation2d(), m_extLeftEncoder.getDistance(), m_extRightEncoder.getDistance());
+    }
+    else {  // internal
+      double leftPosition = m_internalLeftEncoder.getPosition() * m_internalLeftEncoder.getPositionConversionFactor();
+      double rightPosition = m_internalRightEncoder.getPosition() * m_internalRightEncoder.getPositionConversionFactor();
```

Comment:
I believe once you set the conversion factor it is applied already with getPosition(), so this effectively squares the conversion factor. Might want to test to make sure though.